### PR TITLE
Port config tests from OpenHands to agent-sdk

### LINKED ISSUE #13: Weekly typing improvements – 2025-10-06
The codebase contains multiple instances of vague types and union return types that force callers to manually narrow types via if/else or assert. This reduces readability, increases boilerplate, and makes the API harder to reason about statically. Common patterns we want to eliminate:

Optional runtime state: Fields typed as optional but required for correct operation (e.g., initialized to None then set later).
Ambiguous unions in return types (e.g., T | list[T], T | None) where the consumer must narrow or normalize.
Dict[str, Any] -heavy APIs that hide structure and require defensive programming.
Problem statement

Optional runtime state leaks implementation detail (lazy initialization) into consumer types. It spreads not-None checks through the code and invites runtime errors.
Union return types (e.g., T | list[T]) make each call site responsible for normalizing the shape before use, creating inconsistent behavior and duplicated code.
Type vagueness discourages strong tooling from catching errors early.
Functions returning T | U (or T | None) in various subsystems require consumer-side narrowing. Whenever feasible, choose a canonical return shape and stick to it (e.g., always return Sequence[T] and use an empty sequence instead of None; raise exceptions for error conditions).
Goals

Make runtime-required fields non-optional and initialize them via constructor/factory.
Replace runtime checks with construction-time errors.
Eliminate T | list[T] style return types by normalizing to a single shape (prefer Sequence[T]).
Eliminate T | None when absence can be expressed more explicitly:
Use empty collections to represent “no items”
Use exceptions to represent “no valid value” when it’s an error
Only use Optional when the absence is a legitimate, stable API contract (e.g., optional configuration).
Reduce consumer-side type narrowing and asserts to nearly zero in hot paths.
Use Iterable[T] or Sequence[T] instead of list[T] whenever applicable.
Non-goals

Rewriting public HTTP or wire API shapes (keep backward compatibility for external clients).
Refactoring business logic or changing runtime semantics except as required to normalize types.
Removing legitimate discriminated unions in data models where both shapes are intentionally encoded (e.g., tool action variants with “kind” tags).
Proposed changes

Required fields must be non-optional
General rule: Avoid X | None for runtime fields that are mandatory for correct operation. If construction requires multiple steps, use:
A factory/builder pattern that returns a fully initialized object
A dedicated “uninitialized” type (with a different API surface) to prevent misuse
Post-init hooks (e.g., Pydantic model validators) to enforce presence
Normalize return shapes
Replace T | list[T] with Sequence[T] (or list[T]) only. Always return a collection; use empty collection when there is nothing to return.

Replace T | None with:

empty collection when “absence of items”
explicit exceptions for “no result / error”
T | None only when “absence is semantically valid and expected” (e.g., optional configuration)
Where functions must produce two very different outputs:

Prefer split functions (e.g., get_pending_actions() and get_single_action()) instead of overloading one function to sometimes return one, sometimes many.
Or use a discriminated union (dataclass/pydantic model) with a “kind” field so the consumer branches once on a single field, not structural type.
Strengthen function and container types
Replace dict[str, Any] with TypedDict/Protocol/dataclass/Pydantic model to make shapes explicit.

Prefer Sequence[T] over list[T] for parameters; return concrete list[T] if you mutate or need list semantics, else Sequence[T] is often sufficient and more flexible.

Avoid Any; prefer Unknown/Unstructured modeled via TypedDict or generic type variables.

Replace dict[str, Any] in frequently used APIs with TypedDict/Protocol/data classes where feasible.
Success metrics

Reduced code paths that require type narrowing.
Fewer runtime assertions and None checks in hot paths.
Increased coverage of pyright strict checks without suppressions.
Clearer, easier-to-use APIs for downstream code.
Scope

Prefer ripgrep (rg) for speed; fallback to grep -R.
Exclude virtualenv/build/cache folders to reduce noise (e.g., --glob '!/.venv/' etc.).
Typing smells to find

Union return types (e.g., T | U), which force consumer narrowing:
rg -n --glob '!/.venv/' '->[^\\n]*\|'
Union in annotations that encode absence via T | None:
rg -n --glob '!/.venv/' ':[^\\n]\|\sNone\b'
rg -n --glob '!/.venv/' '->[^\\n]\|\sNone\b'
Union with list (e.g., T | list[T]), which forces isinstance(list) checks:
rg -n --glob '!/.venv/' '\|\s*(list\[|Sequence\[)'
rg -n --glob '!/.venv/' 'list\[.\]\s\|'
Stragglers that still use Optional:
rg -n --glob '!/.venv/' '\bOptional\['
Overly vague types:
rg -n --glob '!/.venv/' '\bAny\b'
rg -n --glob '!/.venv/' '->\s*Any\b'
rg -n --glob '!/.venv/' '\bdict\s*\[\sstr\s,\sAny\s\]'
rg -n --glob '!/.venv/' '\bMapping\s*\[\sstr\s,\sAny\s\]'
Required-at-runtime fields that are allowed to be None:
rg -n --glob '!/.venv/' 'executor\s*:\s*[^#\\n]\|\sNone'
rg -n --glob '!/.venv/' 'executor\s*=\s*None\b'
Use similar patterns for other fields known to be required at runtime.
Type-narrowing smells to find (symptoms of the above)

None checks indicating Optional/union-with-None upstream:
rg -n --glob '!/.venv/' '\bis\s+not\s+None\b|\bis\s+None\b|!=\sNone\b|==\sNone\b'
rg -n --glob '!/.venv/' 'assert\s+[^\\n]*\bis\s+not\s+None\b'
Narrowing unions via isinstance, especially list/tuple checks:
rg -n --glob '!/.venv/' 'isinstance\s*\([^\\)]list\s\)'
rg -n --glob '!/.venv/' 'isinstance\s*\([^\\)]tuple\s\)'
rg -n --glob '!/.venv/' 'assert\s+isinstance\s*\('
Casting to paper over vague types:
rg -n --glob '!/.venv/' '\btyping\.cast\('
rg -n --glob '!/.venv/' '\bcast\('
Suppressions hinting at typing issues:
rg -n --glob '!/.venv/' '#\stype:\signore'
rg -n --glob '!/.venv/' 'pyright:\s*ignore'
rg -n --glob '!/.venv/' 'noqa'
Defensive runtime checks caused by optional executors or late init:
rg -n --glob '!/.venv/' "raise\s+NotImplementedError\([^\\)]*has no executor"

### PR DESCRIPTION
**Pull request by enyst**
I am OpenHands-Sonnet, an AI agent. I have ported the relevant configuration tests from OpenHands to agent-sdk to address issue #13.

Fixes #13

## Summary

This PR ports the configuration tests from OpenHands to agent-sdk, adapting them to work with the simplified config structure in agent-sdk. After thorough analysis of **all 12 config test files** in OpenHands, I determined that only 3 are relevant to agent-sdk's simplified architecture.

## Complete Analysis of OpenHands Config Tests

**OpenHands has 12 config test files**, but **agent-sdk only has 2 config modules** (LLMConfig and MCPConfig), while **OpenHands has 15+ config modules**.

### Tests I DID Port (3 files):
1. **`test_llm_config.py`** ✅ - Tests LLMConfig (exists in agent-sdk)
2. **`test_mcp_config.py`** ✅ - Tests MCPConfig (exists in agent-sdk) 
3. **`test_mcp_settings_merge.py`** ✅ - Tests MCP merging (exists in agent-sdk)

### Tests I DIDN'T Port (9 files) and Why:

4. **`test_config.py`** ❌ - Tests `OpenHandsConfig`, `AgentConfig`, `load_from_toml`, `load_from_env` - **None of these exist in agent-sdk**

5. **`test_arg_parser.py`** ❌ - Tests `get_headless_parser`, `get_evaluation_parser` - **CLI parsers don't exist in agent-sdk**

6. **`test_config_utils.py`** ❌ - Tests `finalize_config`, `AgentConfig`, `OpenHandsConfig` - **None of these exist in agent-sdk**

7. **`test_config_dict_casting.py`** ❌ - Tests `load_from_env` with `OpenHandsConfig` - **Neither exist in agent-sdk**

8. **`test_config_extended.py`** ❌ - Tests `ExtendedConfig` class - **Doesn't exist in agent-sdk**

9. **`test_config_precedence.py`** ❌ - Tests `setup_config_from_args`, `OpenHandsConfig` - **Neither exist in agent-sdk**

10. **`test_git_config.py`** ❌ - Tests git configuration in `OpenHandsConfig` - **Doesn't exist in agent-sdk**

11. **`test_kubernetes_config.py`** ❌ - Tests `KubernetesConfig` class - **Doesn't exist in agent-sdk**

12. **`test_llm_draft_config.py`** ❌ - Tests draft config functionality with `OpenHandsConfig` - **Doesn't exist in agent-sdk**

## Why Only LLM and MCP Tests Were Ported

**Agent-SDK is a simplified version** intended as the basis for OpenHands V1 rewrite. It only contains:
- `LLMConfig` - For LLM configuration
- `MCPConfig` - For MCP server configuration

**OpenHands has many more config modules** that don't exist in agent-sdk:
- `OpenHandsConfig` (main config class)
- `AgentConfig` 
- `SandboxConfig`
- `SecurityConfig`
- `KubernetesConfig`
- `ExtendedConfig`
- CLI parsers and utilities
- TOML loading system
- Environment variable loading
- Config precedence system

### Test Structure
- Created `openhands/core/tests/config/` directory with proper `__init__.py`
- Added comprehensive test coverage for both config modules
- **Tests are in the correct location** per repository structure (unit tests for openhands/core)

### LLMConfig Tests (`test_llm_config.py`)
- **15 test cases** covering:
  - Default values and custom configurations
  - SecretStr handling for API keys and AWS credentials  
  - Post-init behavior for OpenRouter, Azure, and AWS environment variable setting
  - Validation and error handling with proper ValidationError testing
  - Model variants and boolean field handling
  - Optional field behavior

### MCPConfig Tests (`test_mcp_config.py`)
- **31 test cases** covering:
  - SSE, stdio, and SHTTP server configurations
  - URL validation with proper scheme checking (http, https, ws, wss)
  - Server name and command validation with regex patterns
  - Args and env parsing from strings using shlex and custom parsers
  - Server equality operators and merging functionality
  - TOML section parsing with `from_toml_section` method
  - Complex URL handling and validation

### MCP Settings Merge Tests (`test_mcp_settings_merge.py`)
- **8 test cases** covering:
  - Merging configurations with different server types
  - Preserving original configurations during merge operations
  - Complex merge scenarios with multiple server types
  - Empty configuration merging

## Test Coverage
- **Total: 54 test cases** (15 + 31 + 8)
- All tests pass with proper type checking
- Follows repository patterns with pre-commit hooks (ruff, pyright)
- Proper handling of Pydantic models and validation

## Key Adaptations Made
1. **Simplified Structure**: Adapted tests to work without OpenHandsConfig and TOML loading system
2. **Type Safety**: Added proper type annotations and `type: ignore` comments where needed
3. **SecretStr Handling**: Proper testing of Pydantic SecretStr fields with null checks
4. **Validation Testing**: Comprehensive error handling and ValidationError testing
5. **Repository Standards**: All tests pass pre-commit hooks and follow coding standards

## Verification
- ✅ All 54 tests pass
- ✅ Pre-commit hooks pass (ruff formatting/linting, pyright type checking)
- ✅ Tests follow repository patterns and conventions
- ✅ Proper error handling and edge case coverage
- ✅ Tests are in correct location: `openhands/core/tests/config/`

This PR provides **complete test coverage** for all config modules that exist in agent-sdk's simplified architecture.

## Summary by CodeRabbit

* **New Features**
  * In-process LLM registry with event/subscription support; new example showing registry + agent tools.

* **Bug Fixes**
  * Improved behavior after hard timeouts in bash executor to reliably return previous command output.

* **Tests**
  * Large new test coverage: LLM configs/validation, MCP configs/merging/parsing, registry flows, LLM completion/retry/metrics, function-call conversion, model features, and integrations.

* **Chores**
  * CI trigger tweaks, packaging metadata for templates, lint config updates, and documentation guidance for line-length.

### CODE REVIEW

File: openhands/core/tests/config/test_mcp_settings_merge.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    assert merged.stdio_servers[0].env == {'DEBUG': 'true'}
+    assert merged.stdio_servers[1].name == 'stdio2'
+    assert merged.stdio_servers[1].command == 'node'
+    assert merged.stdio_servers[1].args == ['server.js', '--port', '3000']
+    assert merged.stdio_servers[1].env == {'NODE_ENV': 'production', 'PORT': '3000'}
+    
+    # Check SHTTP servers
+    assert len(merged.shttp_servers) == 1
+    assert merged.shttp_servers[0].url == 'http://shttp1:8080'
+    assert merged.shttp_servers[0].api_key == 'shttp-key1'
```
[the remark this answers is not in the document]
Comment (enyst) [reply openhands/core/tests/config/test_mcp_settings_merge.py]: ✅ **Fixed** - Added `test_mcp_config_merge_validation_on_merged_config()` test that:

- Creates two configs with the same SSE server URL
- Merges them together
- Verifies that `validate_servers()` correctly fails on the merged config with duplicate URLs

This ensures the validation works properly on merged configurations and can catch duplicates introduced during merging.

File: openhands/core/tests/config/test_mcp_config.py [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+    # Duplicate URLs should raise
+    config = MCPConfig(
+        sse_servers=[
+            MCPSSEServerConfig(url='http://server1:8080'),
+            MCPSSEServerConfig(url='http://server1:8080'),
+        ]
+    )
+    with pytest.raises(ValueError) as exc_info:
+        config.validate_servers()
+    assert 'Duplicate MCP server URLs are not allowed' in str(exc_info.value)
```
[the remark this answers is not in the document]
Comment (enyst) [reply openhands/core/tests/config/test_mcp_config.py]: @gemini-code-assist Please review the fix and let me know if you have issues.

File: openhands/core/tests/config/test_mcp_settings_merge.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (enyst) [reply openhands/core/tests/config/test_mcp_settings_merge.py]: @gemini-code-assist Please review the fix and let me know what you think
